File: code/java/src/main/java/templatemethod/ModernHouseBuilder.java
package templatemethod;

class ModernHouseBuilder extends HouseBuilder {

    @Override
    protected void buildWalls() {
        System.out.println("Building walls with glass and steel.");
    }

    @Override
    protected void installRoof() {
        System.out.println("Installing a flat, green roof.");
    }

    @Override
    protected void doInterior() {
        System.out.println("Doing interior with an open concept design and minimalist furniture.");
    }
}

File: code/typescript/templatemethod/ModernHouseBuilder.ts
namespace TemplateMethod {
    class ModernHouseBuilder extends HouseBuilder {
        protected buildWalls(): void {
            console.log("Building walls with glass and steel.");
        }

        protected installRoof(): void {
            console.log("Installing a flat, green roof.");
        }

        protected doInterior(): void {
            console.log("Doing interior with an open concept design and minimalist furniture.");
        }
    }
}
